GH-116939: Rewrite binarysort()

tim-one opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-116939 -->
* Issue: gh-116939
<!-- /gh-issue-number -->

tim-one committed:
initial

tim-one committed:
Add function comment.

pochmann3 commented on Objects/listobject.c:
```diff
@@ -1665,53 +1674,53 @@ struct s_MergeState {
+                L = M + 1;
+        } while (L < R);
+        assert(L == R);
+        /* a[:L] holds all elements from a[:ok[ <= pivot now, so pivot belongs
```
By now you've used `)`, `]` and `[` on the right side, but I think in all three cases (`s[:ok)` and these two), the right index is exclusive, right?

tim-one replied:
Yes, this is meant to be Python slice notation.. Alas, I'm typing with inadequate light on a new keyboard, and these keys seem endlessly confusable ☹️ .

tim-one committed:
Repair closing braces in comments.

tim-one committed:
Experiments.

Add preprocesor cruft so that you can very easily try either
binary or regular insertion sort. And for binary insertion
sort, switch to a different approach that tries to eliminate the
inner branch by generating conditional move instructions
instead. Alas, Visual Studio will not show me the machine
code generated in a Release build - the compiler and/or linker
apparently optimized it so much it can't relate it to the
source code :-(

tim-one committed:
Remove redundant `return 0;`.

tim-one committed:
Remove unneeded store.

tim-one committed:
And remove needless braces.

tim-one committed:
Comment repair.

tim-one committed:
Modify to account for new MAX_MINRUN #define.

tim-one committed:
Merge branch 'main' into binsort

tim-one committed:
Merge branch 'main' into binsort

pochmann3 commented on Objects/listobject.c:
```diff
@@ -1683,7 +1683,7 @@ binarysort(MergeState *ms, const sortslice *ss, Py_ssize_t n, Py_ssize_t ok)
 
     assert(0 <= ok && ok <= n && 1 <= n && n <= MAX_MINRUN);
-    /* assert s[:ok) is sorted */
+    /* assert s[:ok] is sorted */
```
Is `s` a typo? I'd use `a` throughout. That is, `a[:ok]` here and also in the later invariants replace `0:L` with `a[0:L]` etc.

tim-one replied:
Thanks. It's a typo.

tim-one committed:
Merge branch 'main' into binsort

tim-one committed:
More comment repairs.

tim-one committed:
comment repair

tim-one committed:
Merge branch 'main' into binsort

tim-one committed:
Merge branch 'main' into binsort

tim-one committed:
Merge branch 'main' into binsort

tim-one committed:
Mostly fleshing out comments.
Fiddled the #if's to leave binary insertion the method of choice,
and with the original branch structure. Alternatives are left
in the code, but #if'ed out.

tim-one committed:
Merge branch 'binsort' of https://github.com/tim-one/cpython into binsort
